### PR TITLE
tokens burn feature

### PR DESCRIPTION
**Pull request by TTNguyenDev**
Only burn-token.tokenhub.testnet account has permission to burn tokens

### CODE REVIEW

File: ft/src/lib.rs [hunk 1]
```diff
@@ -31,20 +31,28 @@ near_sdk::setup_alloc!();
 pub struct Contract {
     token: FungibleToken,
     metadata: LazyOption<FungibleTokenMetadata>,
+    token_factory_account: ValidAccountId,
+    burn_token_account: ValidAccountId
```
Comment (lightsea90) [inline ft/src/lib.rs]: `burn_token_account` should be Option since token burn is optional.

**Comment by lightsea90** [comment]
@TTNguyenDev please fix the check failure.

**Comment by TTNguyenDev** [comment]
Okay, the failure comes from the unit test section. I am working on it.

File: ft/src/lib.rs [hunk 2]
```diff
@@ -74,12 +80,35 @@ impl Contract {
...
+            burn_token_account
         };
         this.token.internal_register_account(owner_id.as_ref());
         this.token.internal_deposit(owner_id.as_ref(), total_supply.into());
         this
     }
 
+    pub fn burn_token( 
+        &mut self,
+        account_id: ValidAccountId,
```
Comment (lightsea90) [inline ft/src/lib.rs]: We don't want to burn tokens in _whatever_ account. We only burn token in the account `self.burn_token_account`.

Comment (TTNguyenDev) [reply ft/src/lib.rs]: Okay, let me update

File: ft/src/lib.rs [hunk 3]
```diff
@@ -74,12 +80,35 @@ impl Contract {
...
+        amount: Balance
+    ) {
+
+        if let Some(burn_token_account) = &self.burn_token_account {
```
Comment (lightsea90) [inline ft/src/lib.rs:94]: If `self.burn_token_account == None`, the condition will fail and the code line 103 (internal_withraw) still executed.

File: ft/src/lib.rs (hunk 2 above)
Comment (lightsea90) [reply ft/src/lib.rs]: LGTM

**Comment by vans163** [comment]
Does this follow spec and if not how do the tokens appear in relation to total supply, does total supply go down? https://github.com/near/NEPs/issues/85

**Comment by TTNguyenDev** [comment]
> Does this follow spec and if not how do the tokens appear in relation to total supply, does total supply go down? [near/NEPs#85](https://github.com/near/NEPs/issues/85)

When we do token burn, the total supply goes down.

**Comment by vans163** [comment]
> > Does this follow spec and if not how do the tokens appear in relation to total supply, does total supply go down? [near/NEPs#85](https://github.com/near/NEPs/issues/85)
> 
> When we do token burn, the total supply goes down.

Does burn_token_account equal then to "0000000000000000000000000000000000000000000000000000000000000000" or?

**Comment by TTNguyenDev** [comment]
> > > Does this follow spec and if not how do the tokens appear in relation to total supply, does total supply go down? [near/NEPs#85](https://github.com/near/NEPs/issues/85)
> > 
> > 
> > When we do token burn, the total supply goes down.
> 
> Does burn_token_account equal then to "0000000000000000000000000000000000000000000000000000000000000000" or?

Actually no, you can use any account as a burn_token_account. For our case, we only allow 1 account can burn tokens.

**Comment by vans163** [comment]
I mean how is this a burn when
```
            self.token.internal_withdraw(
                &burn_token_account.to_string(), 
                amount
                );
```
this account can just ft_transfer the tokens later. even if you block the ability to ft_transfer for the burn_token_account, all it takes is 1 WASM bytecode update which will happen in seconds, to remove the block and dump all burned tokens as circulating?

**Comment by TTNguyenDev** [comment]
> I mean how is this a burn when
> 
> ```
>             self.token.internal_withdraw(
>                 &burn_token_account.to_string(), 
>                 amount
>                 );
> ```
> 
> this account can just ft_transfer the tokens later. even if you block the ability to ft_transfer for the burn_token_account, all it takes is 1 WASM bytecode update which will happen in seconds, to remove the block and dump all burned tokens as circulating?

This is the logic behind internal_withdraw
`    pub fn internal_withdraw(&mut self, account_id: &AccountId, amount: Balance) {
        let balance = self.internal_unwrap_balance_of(account_id);
        if let Some(new_balance) = balance.checked_sub(amount) {
            self.accounts.insert(&account_id, &new_balance);
            self.holders.insert(&account_id);
            self.total_supply =
                self.total_supply.checked_sub(amount).expect("Total supply overflow");
        } else {
            env::panic(b"The account doesn't have enough balance");
        }
    }
`

**Comment by TTNguyenDev** [comment]
This is the logic behind internal_withdraw, please take a look on this
Screen Shot 2022-01-30 at 12 45 37

**Comment by lightsea90** [comment]
> I mean how is this a burn when
> 
> ```
>             self.token.internal_withdraw(
>                 &burn_token_account.to_string(), 
>                 amount
>                 );
> ```
> 
> this account can just ft_transfer the tokens later. even if you block the ability to ft_transfer for the burn_token_account, all it takes is 1 WASM bytecode update which will happen in seconds, to remove the block and dump all burned tokens as circulating?

@vans163 This burning mechanism is not like ones in Ethereum or BSC. In ETH/BSC, when you burn, you send tokens to an inaccessible address (usually 0x00...00). The total supply doesn't actually change. Those tokens doesn't disappear. Just no one is able to use them since no one has the private key of that address.

The feature we are proposing here is a real burn. The burned tokens would disappear and total supply would truly decrease. The burn account should usually be managed by the token owner so that only the he/she can manage the total supply.

Does it answer you?

**Comment by vans163** [comment]
> 

Yes so I was wondering about the official way to burn from NEAR Foundation.  Why is a burn_token_account necessary even? Cant you just decrement total supply and remove the tokens off the caller of the burn function?

**Comment by lightsea90** [comment]
> Why is a burn_token_account necessary even? Cant you just decrement total supply and remove the tokens off the caller of the burn function?

So the question is: Should we allow for all accounts to burn? Is it a common practice? @almal do you know about that?
If yes, we can definitely make a change.

**Comment by vans163** [comment]
> > Why is a burn_token_account necessary even? Cant you just decrement total supply and remove the tokens off the caller of the burn function?
> 
> So the question is: Should we allow for all accounts to burn? Is it a common practice? @almal do you know about that? If yes, we can definitely make a change.

Oh I see, you simply limit the ability to burn to `burn_token_account`, right, it might be necessary for your platform to not allow burning except for this account.

**Pull request merged**